packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts

```diff
@@ -0,0 +1,126 @@
+    return new WebSocket(url);
+};
+
+const PING_INTERVAL = 5000;
+const DEBOUNCE_CLOSE_INTERVAL = 10000;
+
+/*
+ * Websocket handling as recommended by RTK QUery see: https://redux-toolkit.js.org/rtk-query/usage/customizing-queries#streaming-data-with-no-initial-request
+ * The websocket is opened after the cacheDataLoaded promise is resolved, and closed after the cacheEntryRemoved promise is resolved maintaining
+ * the connection in between
+ * - It sends a ping message every PING_INTERVAL ms to keep the connection alive
+ * - It debounces the close event to avoid closing the connection every time a 'pong' message is received
+ * - It closes the connection if any websocket error or close event is received
+ * - It invalidates the cache entry with the tag received in the message event
+ */
+const streamMessagesApi = orchestratorApi.injectEndpoints({
+    endpoints: (build) => ({
+        streamMessages: build.query<boolean, void>({
+            queryFn: () => {
+                return { data: false };
+            },
+            async onCacheEntryAdded(
+                _,
+                {
+                    cacheDataLoaded,
+                    cacheEntryRemoved,
+                    dispatch,
+                    getState,
+                    updateCachedData,
+                },
+            ) {
+                const cleanUp = () => {
+                    const message = getToastMessage(
+                        ToastTypes.ERROR,
+                        'Connection to the server was lost. Please refresh the page to reconnect.',
+                        'WebSocket closed',
+                    );
+                    dispatch(addToastMessage(message));
+                    clearInterval(pingInterval);
+                    updateCachedData(() => false);
+                };
+
+                await cacheDataLoaded;
+
+                const state = getState() as RootState;
+                const { orchestratorWebsocketUrl } = state.orchestratorConfig;
+
+                // Starts the websocket
+                const webSocket = getWebSocket(orchestratorWebsocketUrl);
+
+                // Lets the WfoWebsocketStatusBadge know the websocket is connected
+                webSocket.onopen = () => {
+                    updateCachedData(() => true);
+                };
+
+                // Send a ping message every to the websocket server to keep the connection alive
+                const pingInterval = setInterval(() => {
+                    webSocket.send('ping');
+                }, PING_INTERVAL);
+
+                const debounceCloseWebSocket = debounce(() => {
+                    webSocket.close();
+                }, DEBOUNCE_CLOSE_INTERVAL);
+                // Start the debounced function to close the websocket when no 'pong' message is received after DEBOUNCE_CLOSE_INTERVAL
+                debounceCloseWebSocket();
```

Comment:
I guess this is a way, would be nicer to start a timeout after ping is send and clear it after pong is received.

Reply:
I feel this is more stable since you only start the interfal once.